File: measurements/gate_sweep_measurement.py
"""Port of APS2/Measurements/GateSweepMeasurement.m."""

from __future__ import annotations
from dataclasses import dataclass, asdict
from .base_measurement import BaseMeasurement


@dataclass
class GateParams:
    startV: float = -50.0
    maxV: float = 50.0
    points: int = 1001
    process: str = "Sweep_AO"
    output: int = 2


@dataclass
class BiasParams:
    setV: float = 0.1


class GateSweepMeasurement(BaseMeasurement):
    MATLAB_CLASS = "GateSweepMeasurement"

    def __init__(self, bridge, settings: dict,
                 gate: GateParams | None = None,
                 bias: BiasParams | None = None):
        super().__init__(bridge, settings)
        self.gate = gate or GateParams()
        self.bias = bias or BiasParams()

    def create(self):
        super().create()
        self.bridge.assign_fields(self._var + ".Gate", asdict(self.gate))
        self.bridge.assign_fields(self._var + ".Bias", asdict(self.bias))

File: APS2/Measurements/GateSweepMeasurement.m
classdef GateSweepMeasurement < BaseMeasurement
    % GateSweepMeasurement - Performs Gate Sweeps
    
    properties
        Gate struct
        Bias struct
    end
    
    methods
        function obj = GateSweepMeasurement(settings)
            obj@BaseMeasurement(settings);
            
            obj.Settings.type = 'Gatesweep';
            
            % Gate Params
            obj.Gate.startV = -50;
            obj.Gate.maxV = 50;
            obj.Gate.points = 1001;
            obj.Gate.process = 'Sweep_AO'; 
            obj.Gate.output = 2; % Different AO channel
            
            % Bias Params
            obj.Bias.setV = 0.1; % Fixed Bias
        end
        
        function Run(obj)
            fprintf('--- Starting Gate Sweep ---\n');
            
            % 1. Set Fixed Bias First
            % (Assuming Apply_fixed_voltage exists)
            % Apply_fixed_voltage(obj.Settings, obj.Bias);
            
            % 2. Setup ADwin for Gate Sweep
            obj.InitADwinProcess(obj.Gate);
            
            try
                obj.Gate = Run_sweep(obj.Settings, obj.Gate);
                obj.Data.Gate = obj.Gate;
                obj.Data.Bias = obj.Bias;
            catch ME
                fprintf(2, 'Gate Sweep Failed: %s\n', ME.message);
            end
            
            obj.Save();
        end
    end
end